### ISSUE TITLE
Patch from #380 causes t/lru-maintainer.t to fail in Docker for Mac

### ISSUE THREAD

**Issue by shargan** [memcached/memcached#417]
While building memcached from source, `t/lru-maintainer.t` is failing due to the patch introduced in #380.

- Memcached version: 1.5.9 (downloaded from http://www.memcached.org/files/memcached-1.5.9.tar.gz)
- Build environment: Ubuntu Xenial 16.04 docker image (Linux fd785cafeb77 4.9.93-linuxkit-aufs)
- Docker version: Docker for Mac 18.06.0-ce (API version 1.38)
- Mac OS version: macOS High Sierra 10.13.6

To reproduce:
```
root@2cde346f138a:/build/memcached_1.5.9# ./configure
<...snip...>

root@2cde346f138a:/build/memcached_1.5.9# make
<...snip...>

root@2cde346f138a:/build/memcached_1.5.9# make test
<...snip...>
t/lru-maintainer.t .......... 115/226
#   Failed test 'our canary moved to warm'
#   at t/lru-maintainer.t line 80.
#          got: '0'
#     expected: anything else
t/lru-maintainer.t .......... 120/226 # Looks like you failed 1 test of 226.
t/lru-maintainer.t .......... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/226 subtests
<...snip...>

root@7729f355aca3:/build/memcached_1.5.9# ./sizes
Slab Stats  64
Thread stats    -6480
Global stats    168
Settings    248
Item (no cas)   48
Item (cas)  56
Libevent thread 192
Connection  520
----------------------------------------
libevent thread cumulative  6512
Thread stats cumulative     6320
```

1.5.8 does not suffer this issue:
```
root@2cde346f138a:/build/memcached_1.5.9# prove t/lru-maintainer.t
t/lru-maintainer.t .. 99/226
#   Failed test 'our canary moved to warm'
#   at t/lru-maintainer.t line 80.
#          got: '0'
#     expected: anything else
t/lru-maintainer.t .. 120/226 # Looks like you failed 1 test of 226.
t/lru-maintainer.t .. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/226 subtests

Test Summary Report
-------------------
t/lru-maintainer.t (Wstat: 256 Tests: 226 Failed: 1)
  Failed test:  118
  Non-zero exit status: 1
Files=1, Tests=226, 10 wallclock secs ( 0.06 usr  0.01 sys +  0.10 cusr  0.05 csys =  0.22 CPU)
Result: FAIL

root@2cde346f138a:/build/memcached_1.5.9# cd /build/memcached_1.5.8
root@2cde346f138a:/build/memcached_1.5.8# prove t/lru-maintainer.t
t/lru-maintainer.t .. ok
All tests successful.
Files=1, Tests=226,  9 wallclock secs ( 0.06 usr  0.01 sys +  0.08 cusr  0.08 csys =  0.23 CPU)
Result: PASS
```

I can also make 1.5.9 succeed by applying this patch (just a reversal of #380):
```
--- a/t/lru-maintainer.t
+++ b/t/lru-maintainer.t
@@ -56,18 +56,6 @@
         # Items need two fetches to become active
         mem_get_is($sock, "canary", $value);
         mem_get_is($sock, "canary", $value);
-        $stats = mem_stats($sock);
-        # The maintainer thread needs to juggle a bit to actually rescue an
-        # item. If it's slow we could evict after resuming setting.
-        sleep 1;
-        for (0..4) {
-            my $s2 = mem_stats($sock);
-            if ($s2->{lru_maintainer_juggles} - $stats->{lru_maintainer_juggles} < 5) {
-                sleep 1;
-                next;
-            }
-            last;
-        }
     }
     print $sock "set key$key 0 0 66560\r\n$value\r\n";
     is(scalar <$sock>, "STORED\r\n", "stored key$key");
```

**Comment by dormando** [comment]
I have no idea why sleeping longer there would prevent an item from ending up in WARM LRU... wonder if there's clock weirdness allowing the LRU maintainer to decide the tail of WARM is too old :\

can you try this tiny patch, please: https://github.com/dormando/memcached/tree/lru-maint-test2

thanks for the detailed report!

**Comment by shargan** [comment]
That patch resolves the issue for me. I'm afraid I don't know Perl, Prove, or Memcached's internals well enough to even guess at what's going on here. The clock seems to be working as expected--that was recently improved in Docker for Mac--but it's worth noting that this is indeed running on a VM.

**Comment by dormando** [comment]
Something is seriously wrong with thread scheduling/timing under mac :/ I've changed it slightly again to use the total move counter rather than the current item count.

pushed change to 'next', will go out in the next release in a few weeks. thanks again!

**Comment by shargan** [comment]
Just confirming that patch also resolves the issue for me.

Thanks for the quick turnaround.

**Comment by ggardet** [comment]
> can you try this tiny patch, please: https://github.com/dormando/memcached/tree/lru-maint-test2

This patch fixes t/lru-maintainer.t for aarch64, ppc64 and s390x so, @dormando please merge it.

**Comment by dormando** [comment]
@ggardet it is merged, just not released yet. I'll see if I can cut one soon.

**Comment by dormando** [comment]
released now in 1.5.11, sorry for the delay
